zio-http/src/main/scala/zhttp/http/Middleware.scala

```diff
@@ -113,6 +113,30 @@ object Middleware {
   def basicAuth(u: String, p: String): Middleware[Any, Nothing] =
     basicAuth { case (user, password) => (user == u) && (password == p) }
 
+  def addCookieM[R, E](cookie: ZIO[R, E, Cookie]): Middleware[R, E] =
+    patchM((_, _) => cookie.mapBoth(Option(_), c => Patch.addHeader(HttpHeaderNames.SET_COOKIE.toString, c.encode)))
+
+  /**
+   * CSRF middleware : To prevent Cross-site request forgery attacks. This middleware is modeled after the double submit
+   * cookie pattern.
+   * @see
+   *   https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#double-submit-cookie
+   */
+  def csrf(headerName: String, cookieName: String): Middleware[Any, Nothing] = {
+    def getCSRFCookieValue(headers: Headers, cookieName: String): Option[String] =
+      headers.getCookiesDecoded.find(_.name == cookieName).map(_.content)
+    def getCSRFHeaderValue(headers: Headers, headerName: String): Option[String] = headers.getHeaderValue(headerName)
+    Middleware.make((_, _, headers) =>
+      (getCSRFHeaderValue(headers, headerName), getCSRFCookieValue(headers, cookieName)) match {
+        case (Some(headerValue), Some(cookieValue)) =>
+          if (headerValue == cookieValue)
+            true
+          else false
+        case _                                      => false
+      },
+    )((_, _, verified) => if (verified) Patch.empty else Patch.setStatus(Status.FORBIDDEN))
```

Comment:
This might send the body anyway along with Forbidden status code

Reply:
Also add failing test to check body is not sent with any failure status code

Reply:
Added failing test https://github.com/dream11/zio-http/pull/542/files#diff-8f6f8468cde9595e9e0db4f459aee42f03a38213ae7a585f4c54a1cbdb42439fR205

Reply:
Is there a way to change the response body in middleware API ?

Reply:
Fixed